noir-projects/noir-contracts/contracts/token_with_refunds_contract/src/types/token_note.nr

```diff
@@ -72,8 +71,9 @@ impl NoteInterface<TOKEN_NOTE_LEN, TOKEN_NOTE_BYTES_LEN> for TokenNote {
 
 
     fn compute_note_hiding_point(self) -> Point {
-        let (npk_lo, npk_hi) = decompose(self.npk_m_hash);
-        let (random_lo, random_hi) = decompose(self.randomness);
+        // We use the unsafe version because the multi_scalar_mul will constrain the scalars.
+        let npk_m_hash_scalar = from_field_unsafe(self.npk_m_hash);
+        let randomness_scalar = from_field_unsafe(self.randomness);
```

Comment:
Do you know why these two scalars are not constructed the same way as the amoutn scalar?

Reply:
If they all were constructed the same way, then we could have a fn that takes some values and does msm over them generically, instead of having to inline it manually in every place and deal with the unsafe values

Reply:
I would say this is because the npk_m_hash and randomness can span the whole Field so it does not fit only into the lower limb of a scalar. I think Mitch originally wrote this.

@just-mitch do you know if we are really guaranteed that the amount is smaller then lower limb of a grumpkin scalar? Is it the case that it would simply be too huge of an amount of token that you considered it unrealistic to occur?

If this would occur it would brick the contract for the user as MSM would try to constrain the limb and as a result it would fail to generate the proof. Given that we can use from_field_unsafe (which doesn't add constraints) here it's probably better to just call it for amount as well.

Reply:
It'd be fine to constrain token balances to be uint128s - we'd just need to do that consistently. Many DeFi systems use even fewer tokens (e.g. [Balancer uses 112 bits](https://github.com/balancer/balancer-v2-monorepo/blob/36d282374b457dddea828be7884ee0d185db06ba/pkg/vault/contracts/balances/BalanceAllocation.sol#L148)).

Reply:
At any rate this function would likely be incorporated into eg #7585